### PR TITLE
[pigeon] Add method to run pigeon with PigeonOptions

### PR DESCRIPTION
**Pull request by rotorgames**
Add runWithOptions entrypoint to allow external libraries to use the pigeon easier

https://github.com/flutter/flutter/issues/115893

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/pigeon/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 4.2.7
+
+* Add runWithOptions entrypoint to allow external libraries to use the pigeon easier
```
Comment (tarrinneal) [inline packages/pigeon/CHANGELOG.md:3]: Make sure you check out the style guide for changelogs https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style

File: packages/pigeon/lib/pigeon_lib.dart [hunk 2]
```diff
@@ -1327,14 +1327,23 @@ ${_argParser.usage}''';
     }
   }
 
-  /// The 'main' entrypoint used by the command-line tool.  [args] are the
+  // The 'main' entrypoint used by the command-line tool.  [args] are the
```
Comment (tarrinneal) [inline packages/pigeon/lib/pigeon_lib.dart:1330]: probably didn't mean to remove that /

File: packages/pigeon/test/pigeon_lib_test.dart [hunk 3]
```diff
@@ -1210,4 +1210,18 @@ abstract class Api {
     });
     await completer.future;
   });
+
+  test('run with PigeionOptions', () async {
```
Comment (tarrinneal) [inline packages/pigeon/test/pigeon_lib_test.dart:1214]: little typo in pigeon

**Review by tarrinneal — CHANGES_REQUESTED** [review]
a few little nits, I'll let @gaaclarke give final approval.

File: packages/pigeon/CHANGELOG.md (hunk 1 above)
Comment (rotorgames) [reply packages/pigeon/CHANGELOG.md]: done

File: packages/pigeon/lib/pigeon_lib.dart (hunk 2 above)
Comment (rotorgames) [reply packages/pigeon/lib/pigeon_lib.dart]: done

File: packages/pigeon/test/pigeon_lib_test.dart (hunk 3 above)
Comment (rotorgames) [reply packages/pigeon/test/pigeon_lib_test.dart]: done

**Review by tarrinneal — APPROVED** [review]

**Review by gaaclarke — APPROVED** [review]
LGTM, thanks!

**Review by tarrinneal — APPROVED** [review]

**Review by stuartmorgan — COMMENTED** [review]
This will need another version bump due to a conflict with other recent changes, then we land it.

**Comment by rotorgames** [comment]
@stuartmorgan done

**Comment by tarrinneal** [comment]
@rotorgames can you actually merge from main, then add your current changelog update, then send it up one more time?

**Comment by rotorgames** [comment]
@tarrinneal done

**Pull request merged**